src/librbd/deep_copy/ImageCopyRequest.cc

```diff
@@ -177,21 +177,29 @@ int ImageCopyRequest<I>::send_next_object_copy() {
 
   uint64_t ono = m_object_no++;
 
+  uint8_t object_diff_state = object_map::DIFF_STATE_HOLE;
   if (m_object_diff_state.size() > 0) {
     std::set<uint64_t> src_objects;
     map_src_objects(ono, &src_objects);
 
-    bool skip = true;
     for (auto src_ono : src_objects) {
-      if (src_ono >= m_object_diff_state.size() ||
-          m_object_diff_state[src_ono] != object_map::DIFF_STATE_NONE) {
-        skip = false;
-        break;
+      if (src_ono >= m_object_diff_state.size()) {
+        object_diff_state = object_map::DIFF_STATE_DATA_UPDATED;
+      } else {
+        auto state = m_object_diff_state[src_ono];
+        if (state != object_map::DIFF_STATE_HOLE &&
+            ((state == object_map::DIFF_STATE_HOLE_UPDATED &&
+              object_diff_state != object_map::DIFF_STATE_DATA_UPDATED) ||
+             (state == object_map::DIFF_STATE_DATA &&
+              object_diff_state == object_map::DIFF_STATE_HOLE) ||
+             (state == object_map::DIFF_STATE_DATA_UPDATED))) {
```

Comment:
@dillaman nit: was `state != object_map::DIFF_STATE_HOLE` check intentional? Because it looks like it could be skipped, unless it is some sort of optimization (you expect most of time the state will be `HOLE`)?

Reply:
My intention was just to include the four possible cases for clarity. I can remove since it seems to unnecessarily complicate the logic.